File: 2018/14/recipe.hs
{-# LANGUAGE StrictData #-}

import Control.Monad.ST
import Data.STRef
import qualified Data.Vector.Unboxed as V
import qualified Data.Vector.Unboxed.Mutable as MV
import Data.List (findIndex)
import Data.Int (Int8)

type Recipes s = MV.MVector s Int8

data GameState s = GameState { scores :: Recipes s
                             , i1 :: STRef s Int
                             , i2 :: STRef s Int
                             }

initial :: ST s (GameState s)
initial = GameState <$> (V.thaw $ V.fromList [3, 7])
                    <*> newSTRef 0
                    <*> newSTRef 1

stepGame :: GameState s -> ST s (GameState s)
stepGame (GameState scores i1 i2) = do
  s1 <- readSTRef i1 >>= MV.read scores
  s2 <- readSTRef i2 >>= MV.read scores
  let combined = s1 + s2
  scores' <- if combined >= 10
    then do
      s' <- MV.grow scores 2
      MV.write s' (MV.length s' - 2) 1
      MV.write s' (MV.length s' - 1) (combined - 10)
      pure s'
    else do
      s' <- MV.grow scores 1
      MV.write s' (MV.length s' - 1) combined
      pure s'
  modifySTRef i1 $ (`mod` MV.length scores') . (+ (1 + fromIntegral s1))
  modifySTRef i2 $ (`mod` MV.length scores') . (+ (1 + fromIntegral s2))
  pure $ GameState scores' i1 i2
{-# INLINE stepGame #-}

runTilLength :: Int -> GameState s -> ST s (GameState s)
runTilLength n s
  | MV.length (scores s) >= n = pure s
  | otherwise = stepGame s >>= runTilLength n

match :: Recipes s -> V.Vector Int8 -> ST s Bool
match = go 0 where
  go n mvec vec
    | n == MV.length mvec = pure True
    | otherwise = do
        x <- MV.read mvec n
        if x == vec V.! n
          then go (n + 1) mvec vec
          else pure False
{-# INLINE match #-}

solvePt1 :: Int -> ST s [Int8]
solvePt1 n = do
  final <- initial >>= runTilLength (n + 10)
  V.toList <$> (V.freeze $ MV.slice n 10 $ scores final)

solvePt2 :: [Int8] -> ST s Int
solvePt2 digits = do
  let needle = V.fromList $ digits
  s <- initial
  go needle s
  where
    go needle s@(GameState mvec _ _)
      | MV.length mvec < V.length needle = stepGame s >>= go needle
      | otherwise = do
          let lm = MV.length mvec
              ln = V.length needle
          if lm - ln > 0
            then do
              m2 <- match (MV.slice (lm - ln - 1) ln mvec) needle
              if m2
                then pure (lm - ln - 1)
                else do
                  m1 <- match (MV.slice (lm - ln) ln mvec) needle
                  if m1
                    then pure (lm - ln)
                    else stepGame s >>= go needle
            else do
              m1 <- match (MV.slice (lm - ln) ln mvec) needle
              if m1
                then pure (lm - ln)
                else stepGame s >>= go needle

main :: IO ()
main = do
  input <- getLine
  putStrLn $ concatMap show $ runST $ solvePt1 (read input)
  print $ runST $ solvePt2 (read . pure <$> input)

File: 2018/14/recipe.rs
use std::env;

struct GameState {
    scores: Vec<u8>,
    i1: usize,
    i2: usize,
}

impl GameState {
    fn new() -> Self {
        GameState {
            scores: vec![3, 7],
            i1: 0,
            i2: 1,
        }
    }

    #[inline]
    fn step(&mut self) {
        let combined = self.scores[self.i1] + self.scores[self.i2];
        if combined >= 10 {
            self.scores.push(1);
            self.scores.push(combined - 10);
        } else {
            self.scores.push(combined);
        }

        self.i1 = (self.i1 + 1 + self.scores[self.i1] as usize)
            % self.scores.len();
        self.i2 = (self.i2 + 1 + self.scores[self.i2] as usize)
            % self.scores.len();
    }

    fn run_til_len(&mut self, len: usize) {
        while self.scores.len() < len {
            self.step();
        }
    }

    fn run_til_match(&mut self, needle: &[u8]) -> usize {
        while self.scores.len() < needle.len() + 1 {
            self.step();
        }

        loop {
            let ls = self.scores.len();
            let ln = needle.len();
            if &self.scores[(ls - ln - 1)..(ls - 1)] == needle {
                return ls - ln - 1;
            }
            if &self.scores[(ls - ln)..ls] == needle {
                return ls - ln;
            }
            self.step();
        }
    }
}

fn main() {
    let args: Vec<_> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} input", args[0]);
        return;
    }

    // pt 1
    match args[1].parse() {
        Ok(num) => {
            let mut s = GameState::new();
            s.run_til_len(num + 10);
            for i in &s.scores[num..] {
                print!("{}", i);
            }
            println!();
        },

        Err(_) => {
            eprintln!("Invalid input: {}", args[1]);
            return;
        },
    };

    // pt 2
    let needle: Vec<u8> = args[1].chars().map(|c| c as u8 - b'0').collect();
    let mut s = GameState::new();
    let n = s.run_til_match(&needle);
    println!("{}", n);
}
